---
title: "forge_plots.Rmd"
output: html_document
---

```{r libraries}
library(ggplot2)
library(ggrepel)
library(plyr)
library(dplyr)
library(grid)
library(gridExtra)
library(tables)
library(RColorBrewer)
```

```{r set_dir}
# Depends on your system
# setwd('~/git/vis-experiments/Rplot')
```

```{r combined_plot}
# Borrowed from: https://rpubs.com/sjackman/grid_arrange_shared_legend
# Thanks to Shaun Jackman
grid_arrange_shared_legend <- function(...) {
    plots <- list(...)
    g <- ggplotGrob(plots[[1]] + theme(legend.position="bottom"))$grobs
    legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
    lheight <- sum(legend$height)
    grid.arrange(
        do.call(arrangeGrob, lapply(plots, function(x)
            x + theme(legend.position="none"))),
        legend,
        ncol = 1,
        heights = unit.c(unit(1, "npc") - lheight, lheight))
}
```

```{r read_forge}
read_forge <- function (fn, series, blowup_avoidance) {
  tab <- cbind(read.csv(fn, sep='\t'), list('series'=series, 'blowup'=blowup_avoidance))
  tab$incorrect <- tab$Aligned - tab$Overall
  tab$cor_minus_incor <- tab$Overall - tab$incorrect
  tab
}
```

## Fig 1

* Chromosome 19
* Reads simulated from
* Pop cov, hybrid, with blowup avoidance

```{r chr19_sim}
m <- rbind(read_forge('../results/chr9_all_popcov_l100.tsv',        'Population coverage', 'Standard'),
           read_forge('../results/chr9_all_popcov_blowup_l100.tsv', 'Population coverage', 'Blowup avoid'),
           read_forge('../results/chr9_all_amb_l100.tsv',           'Hybrid', 'Standard'),
           read_forge('../results/chr9_all_amb_blowup_l100.tsv',    'Hybrid', 'Blowup avoid'))

plot_forge <- function(m, hap='A') {
  m <- m %>% filter(Hap == hap)
  curve_helper <- function(y_column_name, y_label) {
    m$vertical <- m[[y_column_name]]
    mg <- m %>% dplyr::group_by(series, blowup) %>%
      dplyr::summarise(y=vertical[which(cor_minus_incor == max(cor_minus_incor))[1]],
                       x=Pct[which(cor_minus_incor == max(cor_minus_incor))[1]])
    ggplot(m, aes(x=Pct, y=vertical, color=series, linetype=blowup)) +
      geom_line() +
      geom_point(data=mg, aes(x=x, y=y, fill=series), shape=23, color='black', size=3, alpha=0.5) +
      labs(x='% SNPs', y=y_label) +
      theme_bw() + theme(legend.title=element_blank())
  }
  parametric_helper <- function() {
    mg <- m %>% dplyr::group_by(series, blowup) %>%
      dplyr::summarise(y=Overall[which(cor_minus_incor == max(cor_minus_incor))[1]],
                       x=incorrect[which(cor_minus_incor == max(cor_minus_incor))[1]],
                       pct=Pct[which(cor_minus_incor == max(cor_minus_incor))[1]])
    ggplot(m, aes(x=incorrect, y=Overall, color=series, linetype=blowup)) +
      geom_line() +
      geom_point(data=mg, aes(x=x, y=y, fill=series), shape=23, color='black', size=3, alpha=0.5) +
      geom_text_repel(data=mg, aes(x=x, y=y, label=paste0(pct, '%')), segment.color='grey50', point.padding=0.7) +
      labs(x='% reads incorrect', y='% reads correct') +
      theme_bw() + theme(legend.title=element_blank())
  }
  p_ra <- curve_helper('Aligned', '% reads aligned')
  p_ca <- curve_helper('Correct', '% alignments correct') 
  p_co <- curve_helper('Overall', '% reads correct') 
  p_ci <- parametric_helper()
  grid_arrange_shared_legend(p_ra, p_co, p_ca, p_ci)
}

plot_forge(m)
```

